message-format.md

```diff
@@ -0,0 +1,122 @@
+to the same format below.
+
+The network packets consist of the binary header with the message directly
+appended. There MUST NOT be other data at the end of the packet.
+
+The message MAY be compressed with LZMA, the corresponding header flag signals
+this to the receiver.
+
+If encryption is enabled, a separate encryption header is inserted directly
+after the binary header before the message, a format for this header not yet
+determined
+
+## Binary Header
+```text
+| 0                             | 1                             |
+| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+| MAGIC     | TYPE              | FLAGS                         |
+
+| 2                             | 3                             |
+| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+| RESERVED: 0x00                | RESERVED: 0x00                |
+
+ MAGIC: 0b101
+ TYPE:  0b00000: Standard Message Type
+ FLAGS: 0: compression: 0: None
+                        1: LZMA
+        1: encryption:  0: No
+                        1: Yes
+```
+
+## Definitions
+
+* **REQUIRED**: this entry must be present in every message
+* **REQUIRED, can be omitted if empty**: the sending node must send the
+  information in this entry if it has any. It does not need to include the entry
+  if it would be emtpy
+* **OPTIONAL**: the sending node can decide if it wishes to send this
+  information i.e. it wishes to activate this feature
+
+## Format
+
+```python
+msg = {
+    # the unique router id, may not contain whitespace or any of the brackets:
+    # []()<>{}
+    # REQUIRED
+    "id": "NODE_ID",
+
+    # strictly monotonically increasing, REQUIRED
+    "seq": 12,
+
+    # Announced networks, OPTIONAL
+    "networks": {
+        "1.2.3.4/12": None,
+        # Additional network information, OPTIONAL
+        "2.3.4.5/23": {
+            "retracted": True
+        }
+    },
+
+    # IP addresses of sending interface, at least one is REQUIRED
+    "addr-v4": "123.45.67.89",
+    "addr-v6": "abcd:ef12::1234",
+
+    # all reachable routes, triggers full replacement in receiver when
+    # present, OPTIONAL
+    "routing-data": {
+        # For every policy
+        "POLICY": {
+            # For every reachable node
+            "TO_NODE_ID": {
+                "path": "NODE_ID>[1]>NODE_ID>[2]>NODE_ID",
+            }
+        }
+    },
+
+    # Only partial upgrades, content not yet defined, OPTIONAL
+    "partial-routing-data": {
+        "tbd": "not yet defined"
+    },
+
+    # all nodes referenced in another section, REQUIRED, can be omitted if empty
+    "node-data": {
+        "NODE_ID": {
+            "networks": {
+                "1.2.3.4/12": None,
+                "2.3.4.5/23": {
+                    "retracted": True
+                }
+            }
+        }
+    },
+
+    # all links referenced in "path" above, REQUIRED, can be omitted if empty
+    "link-attributes": {
+        "1": {
+            "bandwidth": 100,
```

Comment:
a general question here: bandwidth is encoded as an json integer. Fine. Probably in the future we see other attributes, not encodeable with integers, probably strings. Is is better to encode values always as strings? Sometimes it is hard because the receiver do not know about the datatype. On the other hand: if the receiver do not know the key (e.g. bandwidth) he know nothing. Thus he cannot interpret the value anyway. What is you suggestion?

Reply:
I think we should encode values as their best representation, ints, floats, string. The policy using these would otherwise have to perform a conversion anyway.

Reply:
ACK